File: subversion/bindings/javahl/src/org/apache/subversion/javahl/types/ExternalItem.java
/**
 * @copyright
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 * @endcopyright
 */

package org.apache.subversion.javahl.types;

import org.apache.subversion.javahl.SubversionException;

/**
 * Describes one external item declaration
 * @since 1.9
 */
public class ExternalItem implements java.io.Serializable
{
    // Update the serialVersionUID when there is a incompatible change made to
    // this class.  See the java documentation for when a change is incompatible.
    // http://java.sun.com/javase/7/docs/platform/serialization/spec/version.html#6678
    private static final long serialVersionUID = 1L;

    /**
     * Create a new external item declaration.
     * <p>
     * <b>Note:</b> If both <code>revision</code> and
     * <code>pegRevision</code> are <code>null</code>, they will be
     * interpreted as {@link Revision#HEAD}. If only one of them is
     * <code>null</code>, it will get the same value as the other
     * revision.
     *
     * @param targetDir See {@link #getTargetDir}
     * @param url See {@link #getUrl}
     * @param revision See {@link #getRevision};
     *     <code>null</code> will be interpreted as <code>pegRevision</code>
     * @param pegRevision See {@link #getPegRevision};
     *     <code>null</code> will be interpreted as {@link Revision#HEAD}
     */
    public ExternalItem(String targetDir, String url,
                        Revision revision, Revision pegRevision)
        throws SubversionException
    {
        this(false, targetDir, url, revision, pegRevision);
        validateRevision(revision, "revision");
        validateRevision(pegRevision, "pegRevision");
    }

    /* This constructor is called directly by the native implementation */
    private ExternalItem(boolean dummy_parameter_to_discriminate_constructors,
                         String targetDir, String url,
                         Revision revision, Revision pegRevision)
    {
        this.targetDir = targetDir;
        this.url = url;
        this.pegRevision = (pegRevision != null ? pegRevision : Revision.HEAD);
        this.revision = (revision != null ? revision : this.pegRevision);
    }

    /**
     * The name of the subdirectory into which this external should be
     * checked out.  This is relative to the parent directory that
     * holds this external item.
     */
    public String getTargetDir()
    {
        return targetDir;
    }

    /**
     * Where to check out from. This is possibly a relative external
     * URL, as allowed in externals definitions, but without the peg
     * revision.
     */
    public String getUrl()
    {
        return url;
    }

    /**
     * What revision to check out. The only valid kinds for this are a
     * numered revision {@link Revision.Number}, a date
     * {@link Revision.DateSpec}, or {@link Revision#HEAD}.
     */
    public Revision getRevision()
    {
        return revision;
    }

    /**
     * The peg revision to use when checking out. The only valid kinds
     * for this are a numered revision {@link Revision.Number}, a date
     * {@link Revision.DateSpec}, or {@link Revision#HEAD}.
     */
    public Revision getPegRevision()
    {
        return pegRevision;
    }

    /**
     * Compare to another external item.
     */
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (!(obj instanceof ExternalItem))
            return false;

        final ExternalItem that = (ExternalItem)obj;
        return (this.targetDir.equals(that.targetDir)
                && this.url.equals(that.url)
                && this.revision.equals(that.revision)
                && this.pegRevision.equals(that.pegRevision));
    }

    /**
     * Return a hash code for this object.
     */
    public int hashCode()
    {
        int hash = 17;
        hash = hash * 33 + (null != targetDir ? targetDir.hashCode() : 0);
        hash = hash * 33 + (null != url ? url.hashCode() : 0);
        hash = hash * 33 + (null != revision ? revision.hashCode() : 0);
        hash = hash * 33 + (null != pegRevision ? pegRevision.hashCode() : 0);
        return hash;
    }

    /* Exception class for failed revision kind validation. */
    private static class BadRevisionKindException extends SubversionException
    {
        // Update the serialVersionUID when there is a incompatible change made to
        // this class.  See the java documentation for when a change is incompatible.
        // http://java.sun.com/javase/7/docs/platform/serialization/spec/version.html#6678
        private static final long serialVersionUID = 1L;

        public BadRevisionKindException(String param)
        {
            super("the '" + param + "' constructor argument" +
                  " must be a date, a number, or Revision.HEAD");
        }
    }

    /* Validates the revision and pegRevision parameters of the ctor. */
    private static void validateRevision(Revision revision, String param)
        throws SubversionException
    {
        if (revision != null
            && revision.getKind() != Revision.Kind.number
            && revision.getKind() != Revision.Kind.date
            && revision.getKind() != Revision.Kind.head)
            throw new BadRevisionKindException(param);
    }

    private String targetDir;
    private String url;
    private Revision revision;
    private Revision pegRevision;
}

File: subversion/bindings/javahl/native/ExternalItem.cpp
/**
 * @copyright
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 * @endcopyright
 */

#include "jniwrapper/jni_stack.hpp"

#include "ExternalItem.hpp"
#include "Revision.h"

namespace JavaHL {

const char* const ExternalItem::m_class_name =
  JAVAHL_CLASS("/types/ExternalItem");

ExternalItem::ClassImpl::ClassImpl(::Java::Env env, jclass cls)
  : ::Java::Object::ClassImpl(env, cls),
    m_mid_ctor(
        env.GetMethodID(cls, "<init>",
                        "(ZLjava/lang/String;Ljava/lang/String;"
                        JAVAHL_ARG("/types/Revision;")
                        JAVAHL_ARG("/types/Revision;")
                        ")V")),
    m_fid_target_dir(env.GetFieldID(cls, "targetDir", "Ljava/lang/String;")),
    m_fid_url(env.GetFieldID(cls, "url", "Ljava/lang/String;")),
    m_fid_revision(env.GetFieldID(cls, "revision",
                                  JAVAHL_ARG("/types/Revision;"))),
    m_fid_peg_revision(env.GetFieldID(cls, "pegRevision",
                                      JAVAHL_ARG("/types/Revision;")))
{}

ExternalItem::ClassImpl::~ClassImpl() {}

namespace {
inline jstring
get_string_field(::Java::Env env, jobject jthis,
                 const ::Java::FieldID& fid)
{
  return jstring(env.GetObjectField(jthis, fid));
}

inline svn_opt_revision_t
get_revision_field(::Java::Env env, jobject jthis,
                   const ::Java::FieldID& fid)
{
  const jobject rev =  env.GetObjectField(jthis, fid);
  return *Revision(rev).revision();
}

inline jobject
make_external_item(::Java::Env env,
                   jclass cls, const ::Java::MethodID& mid_ctor,
                   const char* target_dir,
                   const char* url,
                   const svn_opt_revision_t* revision,
                   const svn_opt_revision_t* peg_revision)
{
  return env.NewObject(cls, mid_ctor,
                       JNI_FALSE,
                       env.NewStringUTF(target_dir),
                       env.NewStringUTF(url),
                       Revision::makeJRevision(*revision),
                       Revision::makeJRevision(*peg_revision));
}
} // anonymous namespace

ExternalItem::ExternalItem(::Java::Env env, jobject jthis)
  : Object(env, ::Java::ClassCache::get_external_item(env), jthis),
    m_target_dir(env, get_string_field(env, jthis, impl().m_fid_target_dir)),
    m_url(env, get_string_field(env, jthis, impl().m_fid_url)),
    m_revision(get_revision_field(env, jthis, impl().m_fid_revision)),
    m_peg_revision(get_revision_field(env, jthis, impl().m_fid_peg_revision))
{}

ExternalItem::ExternalItem(::Java::Env env,
                           const char* target_dir,
                           const char* url,
                           const svn_opt_revision_t* revision,
                           const svn_opt_revision_t* peg_revision)
  : Object(env, ::Java::ClassCache::get_external_item(env)),
    m_target_dir(env, target_dir),
    m_url(env, url),
    m_revision(*revision),
    m_peg_revision(*peg_revision)
{
  set_this(make_external_item(env, get_class(), impl().m_mid_ctor,
                              target_dir, url, revision, peg_revision));
}

svn_wc_external_item2_t*
ExternalItem::get_external_item(SVN::Pool& svnpool) const
{
  svn_wc_external_item2_t* item;
  apr_pool_t* const pool = svnpool.getPool();
  SVN_JAVAHL_CHECK(m_env, svn_wc_external_item2_create(&item, pool));

  item->target_dir = apr_pstrdup(
      pool, Java::String::Contents(m_target_dir).c_str());
  item->url = apr_pstrdup(
      pool, Java::String::Contents(m_url).c_str());
  item->revision = m_revision;
  item->peg_revision = m_peg_revision;
  return item;
}

} // namespace JavaHL


// FIXME: Should be in another source file, but Revision.cpp is
// old-style, so we'll put this implementation here for now.
namespace {
inline jobject get_static_revision(::Java::Env env, jclass cls,
                                   const char* field_name)
{
  return env.GetStaticObjectField(
      cls, env.GetStaticFieldID(cls, field_name,
                                JAVAHL_ARG("/types/Revision;")));
}
} // anonymous namespace

jobject Revision::makeJRevision(const svn_opt_revision_t& rev)
{
  if (rev.kind == svn_opt_revision_number)
      return Revision::makeJRevision(rev.value.number);

  const ::Java::Env env;

  if (rev.kind == svn_opt_revision_date)
    {
      const jclass cls = env.FindClass(
          JAVAHL_CLASS("/types/Revision$DateSpec"));
      return env.NewObject(cls, env.GetMethodID(cls, "<init>", "(J)V"),
                           jlong(rev.value.date / 1000));
    }

  const jclass cls = env.FindClass(JAVAHL_CLASS("/types/Revision"));
  switch (rev.kind)
    {
    case svn_opt_revision_committed:
      return get_static_revision(env, cls, "COMMITTED");

    case svn_opt_revision_previous:
      return get_static_revision(env, cls, "PREVIOUS");

    case svn_opt_revision_base:
      return get_static_revision(env, cls, "BASE");

    case svn_opt_revision_working:
      return get_static_revision(env, cls, "WORKING");

    case svn_opt_revision_head:
      return get_static_revision(env, cls, "HEAD");

    case svn_opt_revision_unspecified:
    default:
      return get_static_revision(env, cls, "UNSPECIFIED");
    }
}
